### PR TITLE
fix binary_crossentropy shape on mps

### LINKED ISSUE #23621: Inconsistent behavior of binary_crossentropy under torch backend
### Description

The Torch backend of `keras.ops.binary_crossentropy` squeezes the final dimension of targets and predictions when MPS is available and that dimension has size 1. It returns the loss without restoring the original shape.

Consequently, identical inputs produce different output shapes depending on MPS availability:

- CPU/CUDA without MPS: `(batch, steps, 1)`
- MPS available: `(batch, steps)`

This breaks code that expects elementwise binary cross-entropy to preserve the input shape.

The condition also checks `torch.backends.mps.is_available()` rather than the tensors' actual device, so it affects CPU tensors on machines where MPS is available.

### Reproduction

Run on a Mac with MPS available:

```python
import os
os.environ["KERAS_BACKEND"] = "torch"

import keras
import torch

assert torch.backends.mps.is_available()

with keras.device("cpu"):
    targets = keras.ops.zeros((3, 4, 1))
    logits = keras.ops.zeros((3, 4, 1))
    loss = keras.ops.binary_crossentropy(
        targets, logits, from_logits=True
    )

    print("Input shape:", targets.shape)
    print("Loss shape:", loss.shape)
    assert tuple(loss.shape) == (3, 4, 1)
```

### Expected behavior

The loss has shape `(3, 4, 1)` consistently across devices.

### Actual behavior

When MPS is available, the loss has shape `(3, 4)`. For example, downstream code using `loss[..., 0]` then produces `(3,)` instead of `(3, 4)`, causes shape mismatches with e.g., sequence masks.

### Relevant implementation

In `keras/src/backend/torch/nn.py`, `binary_crossentropy` explicitly squeezes both inputs before computing the loss, without restoring the dimension afterward. This implementation is present in Keras 3.15.1.

### Suggested fix

Restore the original shape before returning the loss. If the internal squeeze is required only for MPS tensors, derive it from the tensors' actual device rather than global MPS availability.

### PR DESCRIPTION
**Pull request by samudraneel05**
## Description
`keras.ops.binary_crossentropy` under the Torch backend squeezes the trailing dimension when MPS is available, returning `(batch, steps)` instead of `(batch, steps, 1)`. The condition also checks `torch.backends.mps.is_available()` (global) rather than the tensor's actual device, so CPU tensors on MPS-capable machines are also affected.

Root cause: on torch < 2.10, the MPS `binary_cross_entropy` kernel calls
`squeeze()` on its inputs, removing every size-1 dimension, but does not
squeeze `grad_output` ... so backward aborts or returns wrong gradients
(pytorch/pytorch#166746). PR #21816 originally patched a crash in this
backward path (`from_logits=False`)

**Fixes #23621** 

Fixed by having:-
- `from_logits=True`: direct `binary_cross_entropy_with_logits` call, no
  workaround needed.
- `from_logits=False`: compute the loss on 1-D tensors (`reshape(-1)`)
  and reshape the result back. The kernel never sees size-1 dimensions

Verified across backends and devices (took the help of devin to come up with this nifty graphic visualizer):

| Backend | Device | Input shape | Output shape | Status |
|---------|--------|-------------|--------------|--------|
| Torch | CPU | (3, 4, 1) | (3, 4, 1) | fixed |
| Torch | MPS | (3, 4, 1) | (3, 4, 1) | fixed |
| Torch | CPU | (3, 4) | (3, 4) | unchanged |
| Torch | MPS | (3, 4) | (3, 4) | unchanged |
| Torch | CPU | (3, 1) | (3, 1) | fixed |
| Torch | MPS | (3, 1) | (3, 1) | fixed |
| Torch | CPU | (3,) | (3,) | unchanged |
| Torch | MPS | (3,) | (3,) | unchanged |
| Torch | CPU | (2, 3, 4, 1) | (2, 3, 4, 1) | fixed |
| Torch | MPS | (2, 3, 4, 1) | (2, 3, 4, 1) | fixed |
| Torch | MPS | (3, 4, 5) | (3, 4, 5) | unchanged |
| Torch | CPU/MPS | (3, 4, 1) from_logits=False | (3, 4, 1) | fixed |
| JAX | CPU | (3, 4, 1) | (3, 4, 1) | unchanged |
| TensorFlow | CPU | (3, 4, 1) | (3, 4, 1) | unchanged |
| MLX | GPU | (3, 4, 1) | (3, 4, 1) | unchanged |

### Before / after

```python
import os
os.environ["KERAS_BACKEND"] = "torch"
import keras, torch

# On a Mac with MPS available
with keras.device("cpu"):
    targets = keras.ops.zeros((3, 4, 1))
    logits = keras.ops.zeros((3, 4, 1))
    loss = keras.ops.binary_crossentropy(targets, logits, from_logits=True)
```

**Before** (buggy):
```
Input shape: (3, 4, 1)
Loss shape:  (3, 4) 
```

**After** (fixed):
```
Input shape: (3, 4, 1)
Loss shape:  (3, 4, 1)
```

- Loss values match between CPU and MPS (correctness verified).

## Contributor Agreement

**Please review our [PR Contribution Policy](https://github.com/keras-team/keras/issues/23601) and [AI-Assisted Contribution Policy](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] This PR is linked to an issue that has been assigned to me (see `Fixes #xxx` above).
- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed. PRs without a linked, assigned issue will be converted to draft.

### CODE REVIEW

File: keras/src/backend/torch/nn.py [hunk 1]
```diff
@@ -998,16 +998,18 @@ def binary_crossentropy(target, output, from_logits=False):
...
+        target.device.type == "mps"
         and target.ndim > 1
         and output.ndim == target.ndim
         and target.shape[-1] == 1
         and output.shape[-1] == 1
-    ):
+    )
+    if squeeze_trailing:
         target = torch.squeeze(target, -1).contiguous()
         output = torch.squeeze(output, -1).contiguous()
```
[the remark this answers is not in the document]
Comment (samudraneel05) [reply keras/src/backend/torch/nn.py]: point taken, makes sense. testing this simplification and pushing.

File: keras/src/backend/torch/nn.py [hunk 2]
```diff
@@ -998,35 +998,39 @@ def binary_crossentropy(target, output, from_logits=False):
...
             output, target, reduction="none"
         )
     else:
         output = torch.clip(output, backend.epsilon(), 1.0 - backend.epsilon())
-        return tnn.binary_cross_entropy(output, target, reduction="none")
+        loss = tnn.binary_cross_entropy(output, target, reduction="none")
+
+    if squeeze_trailing:
+        loss = loss.unsqueeze(-1)
+    return loss
```
Comment (JyotinderSingh) [inline keras/src/backend/torch/nn.py]: Compute the loss on 1-D tensors instead of squeezing the last dimension. The MPS kernel squeezes every size-1 dimension, so one squeeze covers only part of the problem, and for `(B, T, 1, 1)` it creates the crashing shape. `binary_cross_entropy_with_logits` does not use this kernel, so that branch needs no workaround.

```suggestion
    # By default, PyTorch, does reduction of `sum` over all rows,
    # change reduction to `none` to keep dim
    if from_logits:
        return tnn.binary_cross_entropy_with_logits(
            output, target, reduction="none"
        )
    output = torch.clip(output, backend.epsilon(), 1.0 - backend.epsilon())
    # Before torch 2.10, the MPS `binary_cross_entropy` kernel squeezes every
    # size-1 dimension of its inputs but not of `grad_output`, so the backward
    # pass aborts or returns wrong gradients. Computing the loss on 1-D
    # tensors avoids this on every device.
    # See https://github.com/pytorch/pytorch/issues/166746.
    loss = tnn.binary_cross_entropy(
        output.reshape(-1), target.reshape(-1), reduction="none"
    )
    return loss.reshape(output.shape)
```

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks for picking this up. While checking the tensor's device and restoring the trailing dimension resolves #23621, squeezing a single trailing dimension doesn't fully address the underlying PyTorch bug.

Root cause (pytorch/pytorch#166746): on torch < 2.10, the MPS `binary_cross_entropy` kernel calls `squeeze()` on its input and target, which removes every size-1 dimension. It does not squeeze `grad_output`, and the backward graph multiplies the two. The forward pass is correct. pytorch/pytorch#166786 fixed this in torch 2.10.0. `binary_cross_entropy_with_logits` is a composite op and never uses this kernel.

On PyTorch <= 2.9.1 with MPS (reproduced on 2.2.2; the kernel logic is unchanged through 2.9.1), the current implementation still fails in several scenarios:

- `(B, T, 1, 1)` targets with B != T: the single squeeze creates the `(B, T, 1)` pattern, and the first backward pass aborts the process.
- `(2, 3, 1, 4)` targets: there is no trailing 1 to squeeze, and the backward pass aborts.
- `(B, 1, 1)` targets with `sample_weight`, `class_weight`, or `BinaryFocalCrossentropy`: the loss value is correct, but the gradients are wrong. No error is raised.
- A `(B, 1)` loss and a `(B, 1, 1)` loss in one process (for example, a two-output model): both map to the same cached kernel graph, and the backward pass aborts.

Repro for the wrong gradients (torch <= 2.9.1, MPS):

```python
import torch
import torch.nn.functional as F


def grad(device):
    torch.manual_seed(0)
    p = torch.rand(4, 1, 1).to(device).requires_grad_()
    t = torch.tensor([1.0, 0.0, 1.0, 0.0]).reshape(4, 1, 1).to(device)
    w = torch.tensor([1.0, 2.0, 3.0, 4.0]).reshape(4, 1, 1).to(device)
    # Same squeeze/unsqueeze as this PR.
    loss = F.binary_cross_entropy(
        p.squeeze(-1).contiguous(), t.squeeze(-1).contiguous(), reduction="none"
    ).unsqueeze(-1)
    (loss * w).sum().backward()
    return p.grad.flatten().cpu()


print(grad("cpu"))  # tensor([ -2.0151,   8.6289, -33.9070,   4.6085])
print(grad("mps"))  # tensor([ -2.0151,   4.3145, -11.3023,   1.1521])
```

Flattening the computation to 1-D tensors ensures the kernel never encounters size-1 dimensions, which resolves all four edge cases and produces identical output to PyTorch 2.10. It also keeps execution uniform across devices so standard CI provides coverage. Please see the inline suggestion.

Requested changes:

1. Apply the inline suggestion. It also restores the `from_logits=True` branch to a direct return.
2. Add a float32 regression test: `NNOpsCorrectnessTest.test_binary_crossentropy` currently uses float64 (unsupported on MPS) and passes even if `unsqueeze(-1)` is removed. For example, in keras/src/ops/nn_test.py:

```python
    @parameterized.product(
        shape=[(4, 1), (4, 5, 1), (4, 1, 1), (4, 5, 1, 1), (2, 1, 3)],
        from_logits=[True, False],
    )
    def test_binary_crossentropy_size_one_dims(self, shape, from_logits):
        rng = np.random.default_rng(0)
        target = rng.integers(0, 2, size=shape).astype("float32")
        output = rng.uniform(0.1, 0.9, size=shape).astype("float32")
        result = knn.binary_crossentropy(
            target, output, from_logits=from_logits
        )
        probs = 1.0 / (1.0 + np.exp(-output)) if from_logits else output
        expected = -(target * np.log(probs) + (1 - target) * np.log(1 - probs))
        self.assertEqual(tuple(result.shape), shape)
        self.assertAllClose(result, expected)
```

3. Update the PR description: Clarify that #21816 originally patched an issue with the `binary_cross_entropy` backward pass `(from_logits=False)`, not `binary_cross_entropy_with_logits`.

File: keras/src/backend/torch/nn.py (hunk 2 above)
Comment (samudraneel05) [reply keras/src/backend/torch/nn.py]: Thank you for the detailed comments and code structure. However, I am unable to apply this suggestion as the error message "Applying suggestions on deleted lines is currently not supported" crops up. I shall incorporate these changes and commit while adding you as the coauthor, if that is all right with you.

**Comment by samudraneel05** [comment]
hey @JyotinderSingh, thank you for the detailed review. I had a look through the cited issue in PyTorch and I've incorporated all the requested changes. I've added the float32 regression test, and verified it fails 8/10 cases (of the 5 shapes × 2 from_logits value cases) on the pre-PR implementation on MPS.

**Review by JyotinderSingh — APPROVED** [review]
Thanks for addressing the review! LGTM.

**Review by hertschuh — APPROVED** [review]
Thanks for the fix!
